tools/HeII_input_file_maker.py

```diff
@@ -0,0 +1,292 @@
+        return dataarr
+
+    def _makexHeIIInterp(self):
+        """Produces outfile where columns are z, xHeIII, and number of ionizing photons per nHe produced. Returns an interpolation function."""
+        dataarr = self.xHeIII_quasar(self.z_f, self.z_i)
+        return scipy.interpolate.interp1d(dataarr[0,:], np.log(1e-30+dataarr[1,:]), bounds_error=False, fill_value=0.0)
+
+    def quasar_emissivity_HM12(self, redshift):
+        """Proper emissivity of HeII ionizing photons from Haardt & Madau (2012) (1105.2039.pdf eqn 37)"""
+        enhance_fac=1
+        epsilon_nu = enhance_fac*3.98e24*(1+redshift)**7.68*np.exp(-0.28*redshift)/(np.exp(1.77*redshift) + 26.3) #erg s^-1 MPc^-3 Hz^-1
+        e = epsilon_nu/(self.h_erg_s*self.alpha_q)/(self.mpctocm**3)*4.**(-self.alpha_q)
+        return e
+
+    def quasar_emissivity_K15(self, redshift):
+        """Proper emissivity of HeII ionizing photons from Khaire + (2015)"""
+        epsilon_nu = 10.**(24.6)*(1.+redshift)**8.9 * np.exp(-0.36*redshift)/(np.exp(2.2*redshift)+25.1)  #erg s^-1 MPc^-3 Hz^-1
+        e = epsilon_nu/(self.h_erg_s*self.alpha_q)/(self.mpctocm**3)*4.**(-self.alpha_q)
+        return e
+
+    def quasar_emissivity_Kulkarni19_18(self, redshift):
+        """Proper emissivity of HeII ionizing photons from Kulkarni + (2019) with limiting magnitude -18"""
+        epsilon_nu1450 = 10.**(24.72)*(1.+redshift)**11.42 * np.exp(-2.1*redshift)/(np.exp(1.09*redshift)+38.56)  #erg s^-1 MPc^-3 Hz^-1
+        epsilon_nu = epsilon_nu1450*(912/1450)**0.61
+        e = epsilon_nu/(self.h_erg_s*self.alpha_q)/(self.mpctocm**3)*4.**(-self.alpha_q)
+        return e
+
+    def quasar_emissivity_Kulkarni19_21(self, redshift):
+        """Proper emissivity of HeII ionizing photons from Kulkarni + (2019) with limiting magnitude -21"""
+        #Note the factor of 1+z ^3 compared to Kulkarni 2019: there is a comoving change!
+        epsilon_nu1450 = 10.**(23.91)*(1.+redshift)**11.26 * np.exp(-1.3*redshift)/(np.exp(1.62*redshift)+13.6)  #erg s^-1 MPc^-3 Hz^-1
+        epsilon_nu = epsilon_nu1450*(912/1450)**0.61
+        e = epsilon_nu/(self.h_erg_s*self.alpha_q)/(self.mpctocm**3)*4.**(-self.alpha_q)
+        return e
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser()
+    parser.add_argument('--alphaq', type=float, default=1.7, help='QSO spectral index', required=True)
+    parser.add_argument('--Emax', type=float, default=150, help='Threshold long-mean-free-path photon energy in eV', required=False)
+    parser.add_argument('--cf', type=float, default=3., help='Subgrid clumping factor', required=False)
+    parser.add_argument('--z_i', type=float, default=-1, help='Start redshift of helium reionization', required=False)
+    parser.add_argument('--z_f', type=float, default=-1, help='End redshift of helium reionization', required=False)
+    parser.add_argument('--hist', type=str, default="quasar", help='Type of reionization history', required=True)
```

Comment:
You can declare all allowed choices with https://docs.python.org/3/library/argparse.html#choices

Also, the default z_i, and z_f can probably be written as a dict. Though it is debatable if dict is more readable than if else, shorter and avoids explicit branching certainly.

Reply:
Nice! I did not know about choices.